### PR TITLE
Dispose of HttpResponseMessage after stream is used

### LINKED ISSUE #7248: Image control problem loading from URI
### Steps to Reproduce

1. Create a brand new xamarin forms blank project
2. Add the following code into the mainpage.xaml

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:d="http://xamarin.com/schemas/2014/forms/design"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d"
             x:Class="ImageDemo.MainPage">

    <StackLayout Margin="20,35,20,20">
        <Image Source="https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/Papio_anubis_%28Serengeti%2C_2009%29.jpg/200px-Papio_anubis_%28Serengeti%2C_2009%29.jpg"
               HeightRequest="300" />
    </StackLayout>

</ContentPage>
```

3. Run it

### Expected Behavior
an image would be loaded 

### Actual Behavior
an error is thrown 
Image Loading: Error getting stream for https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/Papio_anubis_(Serengeti%2C_2009).jpg/200px-Papio_anubis_(Serengeti%2C_2009).jpg: System.ObjectDisposedException: Can not access a closed Stream.

### Basic Information

- Version with issue:  
- Last known good version:
Xamarin Forms 4.1.0.673156
Xamarin Essentials  1.2.0

- IDE: VS 2019   16.3.0  Preview 2

- Platform Target Frameworks: 
  - Android:  
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

### Screenshots

[HotReload] (2019-08-22 15:40:48.9): INFO: Xaml Changed for 'MainPage.xaml' in 'ImageDemo', sending to app...
[HotReload] (2019-08-22 15:40:49.7): WARN: Couldn't find correlation for telemetry event
08-22 15:40:49.157 D/Mono    (15293): Loading reference 13 of netstandard.dll asmctx DEFAULT, looking for System.Xml, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
08-22 15:40:49.157 D/Mono    (15293): Assembly Ref addref netstandard[0xe0f85300] -> System.Xml[0xc70b9c00]: 3
08-22 15:40:49.175 D/Mono    (15293): Loading reference 1 of System.Xml.dll asmctx DEFAULT, looking for System, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
08-22 15:40:49.175 D/Mono    (15293): Assembly Ref addref System.Xml[0xc70b9c00] -> System[0xe6d46460]: 8
Xamarin.Android returned no custom HttpClientHandler. Defaulting to System.Net.Http.HttpClientHandler
08-22 15:40:49.501 I/mono-stdout(15293): Xamarin.Android returned no custom HttpClientHandler. Defaulting to System.Net.Http.HttpClientHandler
[HotReload] (2019-08-22 15:40:51.1): INFO: Successfully reloaded: ReloadChange in MainPage.xaml
[0:] Image Loading: Error getting stream for http://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/Papio_anubis_(Serengeti%2C_2009).jpg/200px-Papio_anubis_(Serengeti%2C_2009).jpg: System.ObjectDisposedException: Can not access a closed Stream.
  at System.IO.StreamHelpers.ValidateCopyToArgs (System.IO.Stream source, System.IO.Stream destination, System.Int32 bufferSize) [0x0003a] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at System.IO.MemoryStream.CopyToAsync (System.IO.Stream destination, System.Int32 bufferSize, System.Threading.CancellationToken cancellationToken) [0x00000] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at Xamarin.Forms.UriImageSource.GetStreamAsyncUnchecked (System.String key, System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00310] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:187 
  at Xamarin.Forms.UriImageSource.GetStreamFromCacheAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00140] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:211 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x000f0] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:141 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Threading.CancellationToken userToken) [0x00056] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:80 
[0:] Image loading: Image load failed: System.ObjectDisposedException: Can not access a closed Stream.
  at System.IO.StreamHelpers.ValidateCopyToArgs (System.IO.Stream source, System.IO.Stream destination, System.Int32 bufferSize) [0x0003a] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at System.IO.MemoryStream.CopyToAsync (System.IO.Stream destination, System.Int32 bufferSize, System.Threading.CancellationToken cancellationToken) [0x00000] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at Xamarin.Forms.UriImageSource.GetStreamAsyncUnchecked (System.String key, System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00310] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:187 
  at Xamarin.Forms.UriImageSource.GetStreamFromCacheAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00140] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:211 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x000f0] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:141 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Threading.CancellationToken userToken) [0x000d9] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:91 
  at Xamarin.Forms.Platform.Android.ImageLoaderSourceHandler.LoadImageAsync (Xamarin.Forms.ImageSource imagesource, Android.Content.Context context, System.Threading.CancellationToken cancelationToken) [0x00066] in D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\ImageLoaderSourceHandler.cs:18 
  at Xamarin.Forms.Platform.Android.ResourceManager.GetFormsBitmapAsync (Android.Content.Context context, Xamarin.Forms.ImageSource imageSource, System.Threading.CancellationToken cancellationToken) [0x0006d] in D:\a\1\s\Xamarin.Forms.Platform.Android\ResourceManager.cs:95 
[HotReload] (2019-08-22 15:41:09.8): INFO: Xaml Changed for 'MainPage.xaml' in 'ImageDemo', sending to app...
[HotReload] (2019-08-22 15:41:10.1): WARN: Couldn't find correlation for telemetry event
Xamarin.Android returned no custom HttpClientHandler. Defaulting to System.Net.Http.HttpClientHandler08-22 15:41:09.406 I/mono-stdout(15293): Xamarin.Android returned no custom HttpClientHandler. Defaulting to System.Net.Http.HttpClientHandler

[0:] Image Loading: Error getting stream for https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/Papio_anubis_(Serengeti%2C_2009).jpg/200px-Papio_anubis_(Serengeti%2C_2009).jpg: System.ObjectDisposedException: Can not access a closed Stream.
  at System.IO.StreamHelpers.ValidateCopyToArgs (System.IO.Stream source, System.IO.Stream destination, System.Int32 bufferSize) [0x0003a] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at System.IO.MemoryStream.CopyToAsync (System.IO.Stream destination, System.Int32 bufferSize, System.Threading.CancellationToken cancellationToken) [0x00000] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at Xamarin.Forms.UriImageSource.GetStreamAsyncUnchecked (System.String key, System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00310] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:187 
  at Xamarin.Forms.UriImageSource.GetStreamFromCacheAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00140] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:211 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x000f0] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:141 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Threading.CancellationToken userToken) [0x00056] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:80 
[0:] Image loading: Image load failed: System.ObjectDisposedException: Can not access a closed Stream.
  at System.IO.StreamHelpers.ValidateCopyToArgs (System.IO.Stream source, System.IO.Stream destination, System.Int32 bufferSize) [0x0003a] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at System.IO.MemoryStream.CopyToAsync (System.IO.Stream destination, System.Int32 bufferSize, System.Threading.CancellationToken cancellationToken) [0x00000] in <3e43eda906c6495eb12edfbee9e53c18>:0 
  at Xamarin.Forms.UriImageSource.GetStreamAsyncUnchecked (System.String key, System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00310] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:187 
  at Xamarin.Forms.UriImageSource.GetStreamFromCacheAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x00140] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:211 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Uri uri, System.Threading.CancellationToken cancellationToken) [0x000f0] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:141 
  at Xamarin.Forms.UriImageSource.GetStreamAsync (System.Threading.CancellationToken userToken) [0x000d9] in D:\a\1\s\Xamarin.Forms.Core\UriImageSource.cs:91 
  at Xamarin.Forms.Platform.Android.ImageLoaderSourceHandler.LoadImageAsync (Xamarin.Forms.ImageSource imagesource, Android.Content.Context context, System.Threading.CancellationToken cancelationToken) [0x00066] in D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\ImageLoaderSourceHandler.cs:18 
  at Xamarin.Forms.Platform.Android.ResourceManager.GetFormsBitmapAsync (Android.Content.Context context, Xamarin.Forms.ImageSource imageSource, System.Threading.CancellationToken cancellationToken) [0x0006d] in D:\a\1\s\Xamarin.Forms.Platform.Android\ResourceManager.cs:95 
[HotReload] (2019-08-22 15:41:11.1): INFO: Successfully reloaded: ReloadChange in MainPage.xaml

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

It appears that a race condition with how fast the stream is returned from HttpResponse leads to an image from a stream sometimes being loaded and then sometimes not

The Dispose on HttpContent class will only dispose of the returned stream if the status on the task returned from *ReadAsStreamAsync* is *RantoCompletion*

https://github.com/dotnet/corefx/blob/d3911035f2ba3eb5c44310342cc1d654e42aa316/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L540

So I think in some cases the stream will run to completion before the dispose is called and in other cases it won't

The call to ReadAsStreamAsync returns an invoked task. It doesn't start reading once awaited

I tried moving the await after the dispose explicitly

```C#
using (var client = new HttpClient())
                using (HttpResponseMessage response = await client.GetAsync(uri, cancellationToken))
                {
                    if (!response.IsSuccessStatusCode)
                    {
                        Internals.Log.Warning("HTTP Request", $"Could not retrieve {uri}, status code {response.StatusCode}");
                        return null;
                    }
                    returnValue = response.Content.ReadAsStreamAsync();
                }

                return await returnValue;
```

to see if that would keep it around but it doesn't

We could just not dispose of *HttpResponseMessage* and in theory it would be fine because the GC will eventually dispose of it but since it might be holding on to a large byte array I'd rather it get disposed of quickly. Which is why the solution is a bit more than just not disposing of *HttpResponseMessage*

I also opted out of using the Disposed event on StreamWrapper because I don't want to create a closure 

### Issues Resolved ### 
- fixes #7248 

### Platforms Affected ### 
- Android

### Testing Procedure ###
- UI tests have a few scenarios that use URLs though it seems like it's hit or miss if those catch this one
- the original issue has a few scenarios you can test as well

### PR Checklist ###

- [x] Has automated tests 
- [ ] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/Forms.cs [hunk 1]
```diff
@@ -648,15 +648,18 @@ public double GetNamedSize(NamedSize size, Type targetElementType, bool useOldSi
 
 			public async Task<Stream> GetStreamAsync(Uri uri, CancellationToken cancellationToken)
 			{
-				using (var client = new HttpClient())
-				using (HttpResponseMessage response = await client.GetAsync(uri, cancellationToken))
+				using (var client = new HttpClient())
```
Comment (adrianknight89) [inline Xamarin.Forms.Platform.Android/Forms.cs]: Have you considering using a single static instance of HttpClient across Forms (both on iOS and Android)? In general, it's more efficient to maintain a single active instance than create and dispose multiple instances.

There is some info here that relates to the issue you're fixing: https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

If we maintain a single instance, then maybe all of this work is unnecessary.

Comment (PureWeen) [reply Xamarin.Forms.Platform.Android/Forms.cs]: I tried making HTTPClient static and the same thing happens.

I agree looking into a static instance of HttpClient would be useful but I don't think it's related to this issue.

That blog is talking about issues that come up from making lots of requests and it causing socket exhaustion right? In this case it's a single request that's failing

Also in the comments of that one Oren links to this
http://byterot.blogspot.com/2016/07/singleton-httpclient-dns.html

So it seems like it's kind of a rabbit hole :-)

File: Xamarin.Forms.Platform.Android/Forms.cs [hunk 2]
```diff
@@ -648,15 +648,18 @@ public double GetNamedSize(NamedSize size, Type targetElementType, bool useOldSi
 
 			public async Task<Stream> GetStreamAsync(Uri uri, CancellationToken cancellationToken)
 			{
-				using (var client = new HttpClient())
-				using (HttpResponseMessage response = await client.GetAsync(uri, cancellationToken))
+				using (var client = new HttpClient())				
 				{
+					HttpResponseMessage response = await client.GetAsync(uri, cancellationToken).ConfigureAwait(false);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.Android/Forms.cs]: so, the response is no longer explicitly disposed ?

Comment (PureWeen) [reply Xamarin.Forms.Platform.Android/Forms.cs]: Not as part of this call. It's up to the caller to dispose of the Response if they want it disposed quickly. If they don't dispose of it than the GC will pick it up.

I thought about just reading the response into a MemoryStream and returning that but if you inspect the internals of HttpContent it doesn't always return a MemoryStream so it seemed like a bad idea to just force a full read every time

If we don't like returning the Disposable we could deprecate this Method and force the user to pass in an Action so we control the dispose but I think that would be overkill

Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.Android/Forms.cs]: > It's up to the caller to dispose of the Response

but they don't have any handle to do so

> If they don't dispose of it than the GC will pick it up

eventually, yeah

Comment (PureWeen) [reply Xamarin.Forms.Platform.Android/Forms.cs]: The lifetime of the response is tied to the Stream via the returned StreamWrapper class
https://github.com/xamarin/Xamarin.Forms/pull/7431/files#diff-97e15c2e8d99d55029061029f91ea247R662

https://github.com/xamarin/Xamarin.Forms/pull/7431/files#diff-daba13799e58bccabcbccdc0583fdfa4R75

**Pull request merged**